packages/eslint-plugin/src/rules/no-void-expression.ts

```diff
@@ -0,0 +1,286 @@
+              let newReturnStmtText = `${returnValueText}; return;`;
+              if (returnStmt.parent?.type !== AST_NODE_TYPES.BlockStatement) {
+                // e.g. `if (cond) return console.error();`
+                // add braces if not inside a block
+                newReturnStmtText = `{ ${newReturnStmtText} }`;
+              }
+              return fixer.replaceText(returnStmt, newReturnStmtText);
+            },
+          });
+        }
+
+        // handle generic case
+        if (options.ignoreVoidOperator) {
+          return void context.report({
+            node,
+            messageId: 'invalidVoidExprWrapVoid',
+            suggest: [{ messageId: 'voidExprWrapVoid', fix: wrapVoidFix }],
+          });
+        }
+        context.report({
+          node,
+          messageId: 'invalidVoidExpr',
+        });
+      },
+    };
+
+    /**
+     * Inspects the void expression's ancestors and finds closest invalid one.
+     * By default anything other than an ExpressionStatement is invalid.
+     * Parent expressions which can be used for their short-circuiting behavior
+     * are ignored and their parents are checked instead.
+     * @param node The void expression node to check.
+     * @returns Invalid ancestor node if it was found. `null` otherwise.
+     */
+    function findInvalidAncestor(node: TSESTree.Node): TSESTree.Node | null {
+      const { parent } = node;
+
+      if (parent == null) {
+        // if there is no parent then let's say it's valid
+        return null;
+      }
+
+      if (parent.type === AST_NODE_TYPES.ExpressionStatement) {
+        // e.g. `{ console.log("foo"); }`
+        // this is always valid
+        return null;
+      }
+
+      if (parent.type === AST_NODE_TYPES.LogicalExpression) {
+        if (parent.right === node) {
+          // e.g. `x && console.log(x)`
+          // this is valid only if the next ancestor is valid
+          return findInvalidAncestor(parent);
+        }
+      }
+
+      if (parent.type === AST_NODE_TYPES.ConditionalExpression) {
+        if (parent.consequent === node || parent.alternate === node) {
+          // e.g. `cond ? console.log(true) : console.log(false)`
+          // this is valid only if the next ancestor is valid
+          return findInvalidAncestor(parent);
+        }
+      }
+
+      if (parent.type === AST_NODE_TYPES.ArrowFunctionExpression) {
+        // e.g. `() => console.log("foo")`
+        // this is valid with an appropriate option
+        if (options.ignoreArrowShorthand) {
+          return null;
+        }
+      }
+
+      if (parent.type === AST_NODE_TYPES.UnaryExpression) {
+        if (parent.operator === 'void') {
+          // e.g. `void console.log("foo")`
+          // this is valid with an appropriate option
+          if (options.ignoreVoidOperator) {
+            return null;
+          }
+        }
+      }
+
+      // any other parent is invalid
+      return parent;
```

Comment:
I personally think that this pattern is clearer as a switch:

```suggestion
    switch (parent.type) {
      case AST_NODE_TYPES.ExpressionStatement:
        // e.g. `{ console.log("foo"); }`
        // this is always valid
        return null;

      case  AST_NODE_TYPES.LogicalExpression:
        if (parent.right === node) {
          // e.g. `x && console.log(x)`
          // this is valid only if the next ancestor is valid
          return findInvalidAncestor(parent);
        }
        return parent;
     

      case AST_NODE_TYPES.ConditionalExpression:
        if (parent.consequent === node || parent.alternate === node) {
          // e.g. `cond ? console.log(true) : console.log(false)`
          // this is valid only if the next ancestor is valid
          return findInvalidAncestor(parent);
        }
        return parent;
        
      case AST_NODE_TYPES.ArrowFunctionExpression:
        // e.g. `() => console.log("foo")`
        // this is valid with an appropriate option
        if (options.ignoreArrowShorthand) {
          return null;
        }
        return parent;

      case AST_NODE_TYPES.UnaryExpression:
        if (parent.operator === 'void') {
          // e.g. `void console.log("foo")`
          // this is valid with an appropriate option
          if (options.ignoreVoidOperator) {
            return null;
          }
        }
        return parent;

      default:
        // any other parent is invalid
        return parent;
    }
```

Reply:
I don't like how you have to repeat `return parent;` every time and only add the comment to the last one, even though this comment applies to every one.

I think it's more obvious when it's used only once at the end.